src/decoder/mod.rs

```diff
@@ -187,6 +189,12 @@ impl<R: Read> Decoder<R> {
     pub fn set_transformations(&mut self, transform: Transformations) {
         self.transform = transform;
     }
+
+    /// Set the custom streaming decoder
+    pub fn set_streaming_decoder(&mut self, streaming_decoder: StreamingDecoder) {
+        self.streaming_decoder = streaming_decoder;
+    }
```

Comment:
I see the benefits of this idea, it is a good decomposition in principle. However, it's may not be as simple as it appears and is certainly more complex than having its own `set_ignore_text_chunk`.

* Can you provide the usage with `set_ignore_text_chunk` as an example? This is a somewhat implicit way of performing the desired operation and I fear that some users may not understand exactly how to make use of it.

* If the stream decoder has been used before, is this intended to work correctly? What exactly _is_ correct? Should it be possible to decoder the header with an instance then pass it to `Decoder` and expect it to continue? Or should the `Decoder` always restart at a new `IHDR`?

Reply:
You are right, I looked at StreamingDecoder and realized that we should not use a low-level interface(StreamingDecoder) out of Decoder, it will confuse the other users. However, I think ignore_text_chunk should be a Decoder configuration that is passed to StreamingDecoder when inner StreamingDecoder is used. but in the current interface, for example, read_info, we lack to set StreamingDecoder to ignore_text_chunk.

this set_streaming_decoder may change to set_decoder_config and pass the setting to streaming_decoder. and I will give a commit soon.

> The linked issue is about incorrect errors during decoding of images with image-rs/png. The specific images in question had been encoded by Photoshop, as far as I know, and it was also related to a specific text chunk that Photoshop apparently inserts.

Yes, it may have come from Photoshop, I found the error image has some jpeg info written in the iTXT chunk. Looks like it did some matting.

Reply:
In order to avoid duplication but also allow configuration `StreamingDecoder` in `read_info`, what are your thoughts about introducing the functionality like so:

```rust
/// Private, for now. Mostly because naming is hard and I'm unsure about the scope of its responsibilities.
/// Just want to clarify which fields the Decoder shares with StreamingDecoder.
#[derive(Clone, Copy)]
// Actually, Copy may be too restrictive in the future. So another reason not to expose this yet.
struct ChunkHandling {
    ignore_txt_chunk: bool,
    // In a separate PR, maybe we should move `limit: Limit` into this?
}

pub struct Decoder<R: Read> {
    …,
    chunks: ChunkHandling,
}

struct StreamingDecoder {
    …,
    pub(crate) chunks: ChunkHandling,
}

impl<R: Read> Decoder<R> {
    pub fn read_info(self) -> Result<Reader<R>, DecodingError> {
        let mut stream =  StreamingDecoder::new();
        stream.chunks  = self.chunks;
        let mut reader = Reader::new(self.r, stream, self.transform, self.limits);
        reader.init()?;
    }

   pub fn set_ignore_text_chunk(&mut self, ignore_text_chunk: bool) {
        self.chunks.ignore_text_chunk = ignore_text_chunk;
    }
}
```

This way the details are hidden (for now) while we iterate over the exact options and control that we want to allow the caller to customize in chunk handling.

Reply:
yes, that is basically how I did it. 

and this is the image I mentioned, you can look into it if you're interested.

* https://img.alicdn.com/i3/6000000005544/O1CN01lz0UEm1qpCWNP6YW8_!!6000000005544-2-mia.png

Reply:
Your image seems to have been affected, and #337 fixed it. Locally,

```bash
$ cargo run --example pngcheck -- /tmp/O1CN01lz0UEm1qpCWNP6YW8_!!6000000005544-2-mia.png
OK: /tmp/O1CN01lz0UEm1qpCWNP6YW8_!!6000000005544-2-mia.png (2362x2362, 8-bit RGB+alpha, non-interlaced, 92.6%)
```